samples/wire/README.md

```diff
@@ -0,0 +1,317 @@
+
+``` go
+func NewGreeter(m Message) Greeter {
+    var grumpy bool
+    if time.Now().Unix()%2 == 0 {
+        grumpy = true
+    }
+    return Greeter{Message: m, Grumpy: grumpy}
+}
+```
+
+We have added a `Grumpy` struct field to `Greeter` and if the invocation time
+of the initializer is an even number of seconds since the epoch, we will create
+a grumpy greeter instead of a friendly one.
+
+The `Greet` method then becomes:
+
+``` go
+func (g Greeter) Greet() Message {
+    if g.Grumpy {
+        return Message("Go away!")
+    }
+    return g.Message
+}
+```
+
+Now you see how a grumpy `Greeter` is no good for an `Event`. So `NewEvent`
+may fail. Our `main` must now take into account that `InitializeEvent` may in
+fact fail:
+
+``` go
+func main() {
+    e, err := InitializeEvent()
+    if err != nil {
+        fmt.Printf("failed to create event: %s\n", err)
+        os.Exit(2)
+    }
+    e.Start()
+}
+```
+
+We also need to update `InitializeEvent` to add an `error` type to the return value:
+
+``` go
+// wire.go
+
+func InitializeEvent() (Event, error) {
+    wire.Build(NewEvent, NewGreeter, NewMessage)
+    return Event{}, nil
+}
+```
+
+With the setup complete, we are ready to invoke the `wire` command again. Note,
+that after running `wire` once to produce a `wire_gen.go` file, we may also use
+`go generate`. Having run the command, our `wire_gen.go` file looks like
+this:
+
+``` go
+// wire_gen.go
+
+func InitializeEvent() (Event, error) {
+    message := NewMessage()
+    greeter := NewGreeter(message)
+    event, err := NewEvent(greeter)
+    if err != nil {
+        return Event{}, err
+    }
+    return event, nil
+}
+```
+
+Wire has detected that the `NewEvent` provider may fail and has done the right
+thing inside the generated code: it checks the error and returns early if one
+is present.
+
+As another improvement, let's look at how Wire generates code based on the
+signature of the injector. Presently, we have hard-coded the message inside
+`NewMessage`. In practice, it's much nicer to allow callers to change that
+message however they see fit. So let's change `InitializeEvent` to look like this:
+
+``` go
+func InitializeEvent(phrase string) (Event, error) {
+    wire.Build(NewEvent, NewGreeter, NewMessage)
+    return Event{}, nil
+}
+```
+
+Now `InitializeEvent` allows callers to pass in the `phrase` for a `Greeter` to
+use. We also add a `phrase` argument to `NewMessage`:
+
+``` go
+func NewMessage(phrase string) Message {
+    return Message(phrase)
+}
+```
+
+After we run `wire` again, we will see that the tool has generated an
+initializer which passes the `phrase` value as a `Message` into `Greeter`.
+Neat!
+
+``` go
+// wire_gen.go
+
+func InitializeEvent(phrase string) (Event, error) {
+    message := NewMessage(phrase)
+    greeter := NewGreeter(message)
+    event, err := NewEvent(greeter)
+    if err != nil {
+        return Event{}, err
+    }
+    return event, nil
+}
+```
```

Comment:
One last thing that would be nice to showcase here: Wire's error messages. It doesn't have to be a long thing, but you could instruct the reader to remove one of the arguments to `wire.Build` (perhaps the `NewGreeter` call) and show that Wire points out what is missing. Another nice one to show would be to add in a superfluous function and show that Wire tells you that it is unused.

For someone who is first using Wire, I don't want them to be afraid of error messages. Much like Go compiler error messages, the intent is not to criticize, but to aid the programmer into writing the next function. If they see this being done with intent in the sample (a controlled environment), then it's more likely that when they come across it in practice that they will know that Wire is helping them know what to do next.

Reply:
Great point. I'll add your two suggestions.